### PR TITLE
Deep space

### PR DESCRIPTION
**Pull request by agurfinkel**
Fixing and improving iuc extraction from proofs

### CODE REVIEW

File: src/muz/spacer/spacer_unsat_core_learner.cpp [hunk 1]
```diff
@@ -36,28 +36,27 @@ void unsat_core_learner::register_plugin(unsat_core_plugin* plugin) {
...
-        // if the node mixes A-reasoning and B-reasoning and contains non-closed premises
-        if (m_pr.is_a_marked(currentNode) && 
-            m_pr.is_b_marked(currentNode) && 
-            !m_closed.is_marked(currentNode)) {
-            compute_partial_core(currentNode); // then we need to compute a partial core
+        // we have now collected all necessary information,
+        // so we can visit the node
+        // if the node mixes A-reasoning and B-reasoning
+        // and contains non-closed premises
+        if (!done) {
```
Comment (NikolajBjorner) [inline src/muz/spacer/spacer_unsat_core_learner.cpp]: always true

File: src/muz/spacer/spacer_unsat_core_plugin.cpp [hunk 2]
```diff
@@ -134,34 +133,35 @@ namespace spacer {
...
+                  for (unsigned i = 0; i < m.get_num_parents(step); ++i) {
+                      proof * prem = m.get_parent(step, i);
+                      rational coef = params[i].get_rational();
+                      bool b_pure = m_ctx.is_b_pure (prem);
+                      tout << (b_pure?"B":"A") << " " << coef << " " << mk_pp(m.get_fact(prem), m) << "\n";
+                  }
+                );
+
+            bool done = true;
+
```
Comment (NikolajBjorner) [inline src/muz/spacer/spacer_unsat_core_plugin.cpp]: there is an iterator for that. It is called get_parents(step), so you say
for (proof * premise : get_parents(step)) {…. }

Comment (NikolajBjorner) [reply src/muz/spacer/spacer_unsat_core_plugin.cpp]: oh, well, unless you need to access params[i]

File: src/muz/spacer/spacer_unsat_core_learner.cpp (hunk 1 above)
Comment (agurfinkel) [reply src/muz/spacer/spacer_unsat_core_learner.cpp]: @NikolajBjorner why? `done` is true if all parents are `done`. This is entered only if there is an open parent that has not been interpolated over.

**Pull request merged**